### PR TITLE
refactor(harmony): clean up parser signatures, fix delta accumulation, and GptOss tests

### PR DESCRIPTION
**Pull request by CatherineSue**
### Problem

Fixes: #575

1. `matched_stop` was threaded through the parser's `parse_complete`/`finalize` signatures and stored in `HarmonyChannelOutput`, but callers already have access to it directly from the proto wrapper. This was unnecessary passthrough.

2. In `parse_chunk`, when multiple tokens in a single chunk produce deltas for the same channel, only the last delta survived (overwrite instead of accumulate). Additionally, `has_delta` was tracked via a mutable flag set in multiple places instead of being derived from actual output.

3. The gpt-oss (Harmony) stop sequence tests asserted that stop sequences are trimmed from output, but Harmony's `StreamableParser` does not expose per-channel token IDs, so token-level stop trimming (like `StopSequenceDecoder` does for the regular path) is not possible.

### Solution

1. Remove `matched_stop` from `parse_complete`/`finalize` signatures and `HarmonyChannelOutput`. Callers now read `matched_stop` directly from the proto wrapper.

2. Accumulate deltas across tokens in `parse_chunk` using `get_or_insert_with(String::new).push_str()` and derive `has_delta` from whether any channel produced output.

3. Add `STOP_SEQUENCE_TRIMMED` class flag to E2E tests. Regular backends assert stop sequences are absent; Harmony asserts the stop sequence is a suffix (present but at the end).

## Changes

- `parser.rs`: Remove `matched_stop` param from `parse_complete()`/`finalize()`; fix multi-token delta accumulation in `parse_chunk()` with `get_or_insert_with`; derive `has_delta` from output
- `processor.rs`: Simplify `parse_complete()` calls to 2 args; remove unused `matched_stop` variable in Responses path
- `streaming.rs`: Remove `matched_stop` from `finalize()` calls; avoid redundant `finalize()` call by reusing saved analysis; use direct move instead of `clone_from` for `accumulated_tool_calls`
- `types.rs`: Remove `matched_stop` field from `HarmonyChannelOutput`; remove unused `serde_json::Value` import
- `test_openai_server.py`: Add `STOP_SEQUENCE_TRIMMED` flag (`True` for regular, `False` for Harmony)

## Test Plan

- `cargo build` passes
- `cargo clippy -- -D warnings` passes
- E2E stop sequence tests adapt via `STOP_SEQUENCE_TRIMMED` flag

### CODE REVIEW

File: model_gateway/src/routers/grpc/harmony/streaming.rs [hunk 1]
```diff
@@ -301,8 +475,22 @@ impl HarmonyStreamingProcessor {
...
+                .map(|s| s.to_vec())
+                .unwrap_or_default()
+        };
+
         // Phase 2: Process decode stream (same as single stream)
         let mut parsers: HashMap<u32, HarmonyParserAdapter> = HashMap::new();
+        let mut stop_buffers: HashMap<u32, StopSequenceBuffer> = HashMap::new();
         let mut is_firsts: HashMap<u32, bool> = HashMap::new();
         let mut finish_reasons: HashMap<u32, Option<String>> = HashMap::new();
         let mut matched_stops: HashMap<u32, Option<serde_json::Value>> = HashMap::new();
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply model_gateway/src/routers/grpc/harmony/streaming.rs]: Valid observation, but this duplication is pre-existing and not introduced by this PR. Extracting a shared `process_decode_phase` helper would be a worthwhile follow-up refactor, but it's out of scope here to keep the PR focused.

File: model_gateway/src/routers/grpc/harmony/parser.rs [hunk 2]
```diff
@@ -93,6 +93,59 @@ impl HarmonyParserAdapter {
...
+                        let new_len = text.len() - stop_str.len();
+                        text.truncate(new_len);
+                    }
+                }
+                if final_text.ends_with(stop_str) {
+                    let new_len = final_text.len() - stop_str.len();
+                    final_text.truncate(new_len);
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply model_gateway/src/routers/grpc/harmony/parser.rs]: Good suggestion — addressed in 5bcc6355. Extracted a `strip_suffix` closure to DRY up the match arms.

File: e2e_test/chat_completions/test_openai_server.py [hunk 3]
```diff
@@ -397,9 +400,14 @@ def test_stop_sequences_stream(self, setup_backend, smg):
...
+        # Collect all content (fall back to reasoning_content for models like Harmony)
+        def _delta_text(delta):
+            return delta.content or getattr(delta, "reasoning_content", "") or ""
+
         content = "".join(
-            c.choices[0].delta.content for c in chunks if c.choices and c.choices[0].delta.content
+            _delta_text(c.choices[0].delta)
+            for c in chunks
+            if c.choices and _delta_text(c.choices[0].delta)
         )
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply e2e_test/chat_completions/test_openai_server.py:422]: Good call — addressed in 5bcc6355. Promoted `_delta_text` to a `@staticmethod` on the test class.

File: model_gateway/src/routers/grpc/harmony/parser.rs [hunk 4]
```diff
@@ -93,6 +93,59 @@ impl HarmonyParserAdapter {
...
+                    if text.ends_with(stop_str) {
+                        let new_len = text.len() - stop_str.len();
+                        text.truncate(new_len);
+                    }
+                }
+                if final_text.ends_with(stop_str) {
+                    let new_len = final_text.len() - stop_str.len();
+                    final_text.truncate(new_len);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply model_gateway/src/routers/grpc/harmony/parser.rs]: Agreed — addressed in 5bcc6355. Added `tracing::trace!` with the channel value in the unknown-channel fallback arm.

File: model_gateway/src/routers/grpc/harmony/streaming.rs [hunk 5]
```diff
@@ -44,6 +44,116 @@ use crate::{
...
+                .unwrap_or(buf.len());
+            let safe = buf[..keep_from].to_string();
+            buf.drain(..keep_from);
+            if !safe.is_empty() {
+                return Some(safe);
+            }
+        }
+
+        None
+    }
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply model_gateway/src/routers/grpc/harmony/streaming.rs]: Valid point about test coverage for UTF-8 edge cases in `StopSequenceBuffer::buffer()`. However, this code was introduced in a prior commit (651a13f7) and isn't changed in this PR. We'll track this as a follow-up.

File: model_gateway/src/routers/grpc/harmony/streaming.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (CatherineSue) [reply model_gateway/src/routers/grpc/harmony/streaming.rs]: Resolved.

File: e2e_test/chat_completions/test_openai_server.py [hunk 6]
```diff
@@ -558,6 +590,12 @@ class TestChatCompletionGptOss(TestChatCompletion):
     with OSS models. Logprobs are supported via Harmony's built-in tokenizer.
     """
 
+    # Harmony channel markers add ~10 special tokens
+    STREAMING_TOKEN_TOLERANCE = 10
+
+    # Harmony doesn't trim stop sequences (detokenization is not channel-aware)
+    STOP_SEQUENCE_TRIMMED = False
+
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply e2e_test/chat_completions/test_openai_server.py:598]: Harmony doesn't support trim.
